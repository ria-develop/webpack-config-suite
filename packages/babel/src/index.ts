export * from './JsxExtensions';
export * from './HotModuleReplacementConfig';
export * from './BabelReactConfig';
export * from './BabelConfig';
