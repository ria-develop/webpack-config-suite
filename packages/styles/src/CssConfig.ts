import {
  BaseConfig,
  getLoader,
  ExtractCssLoaderOptions,
  StyleLoaderOptions,
  PostCssLoaderOptions,
  CssLoaderOptions,
  RuleOptimizationAwareConfig
} from '@webpack-config-suite/core';
import {CssExtension, CssPlugins} from './index';
import {RuleSetUseItem} from 'webpack';
import MiniCssExtractPlugin from 'mini-css-extract-plugin';

export class CssConfig extends RuleOptimizationAwareConfig {
  public get composed(): typeof BaseConfig[] {
    return [CssExtension, CssPlugins];
  }

  protected get test(): RegExp | string {
    return /\.css/;
  }

  protected get cssLoaderOptions(): CssLoaderOptions {
    return {sourceMap: this.isDev};
  }

  protected get cssLoader(): RuleSetUseItem {
    return getLoader('css-loader', this.cssLoaderOptions);
  }

  protected get styleLoaderOptions(): StyleLoaderOptions {
    return null;
  }

  protected get styleLoader(): RuleSetUseItem {
    return getLoader('style-loader', this.styleLoaderOptions);
  }

  protected get postCssLoaderOptions(): PostCssLoaderOptions {
    return {
      postcssOptions: {
        plugins: [['autoprefixer']]
      }
    };
  }

  protected get postCssLoader(): RuleSetUseItem {
    return getLoader('postcss-loader', this.postCssLoaderOptions);
  }

  protected get extractCssLoaderOptions(): ExtractCssLoaderOptions {
    return {
      esModule: false,
      // only enable hot in development
      //hmr: args.mode === 'development' && args.argv.hot,
      hmr: this.isHot,
      // if hmr does not work, this is a forceful method.
      reloadAll: true
    };
  }

  protected get extractCssLoader(): RuleSetUseItem {
    return getLoader(MiniCssExtractPlugin.loader, this.extractCssLoaderOptions);
  }

  /**
   *
   * @returns {[]}
   */
  protected get use(): RuleSetUseItem[] {
    return [this.styleLoader, this.extractCssLoader, this.cssLoader, this.postCssLoader];
  }
}
