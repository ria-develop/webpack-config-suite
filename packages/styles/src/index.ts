export * from './CssConfig';
export * from './CssExtension';
export * from './CssPlugins';
export * from './LessConfig';
export * from './LessExtension';
