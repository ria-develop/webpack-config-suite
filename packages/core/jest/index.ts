export {TestCases} from '../../../jest';
import {processConfig, WebpackCliOptions, What} from '../src';
import {Configuration} from 'webpack';

export function testProcessConfig(
  Entity: What,
  config: Configuration,
  env: unknown = undefined,
  argv: WebpackCliOptions = {}
): void {
  const composed = processConfig(Entity, config, env, argv);
  if (composed) {
    composed.forEach((ComposedConfig) => testProcessConfig(ComposedConfig, config, env, argv));
  }
}
