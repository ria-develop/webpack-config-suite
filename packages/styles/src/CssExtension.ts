import {ExtensionsAwareConfig} from '@webpack-config-suite/core';

export class CssExtension extends ExtensionsAwareConfig {
  protected get extensions(): string[] {
    return ['.css'];
  }
}
