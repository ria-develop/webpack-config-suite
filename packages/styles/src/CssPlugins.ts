import ExtractCssPlugin from 'mini-css-extract-plugin';
import {PluginsAwareConfig, MiniCssExtractPluginOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';

export class CssPlugins extends PluginsAwareConfig {
  protected get extractCssPluginOptions(): MiniCssExtractPluginOptions {
    return undefined;
  }

  protected get extractCssPlugin(): ExtractCssPlugin {
    return new ExtractCssPlugin(this.extractCssPluginOptions);
  }

  get plugins(): Configuration['plugins'] | undefined {
    return [this.extractCssPlugin];
  }
}
