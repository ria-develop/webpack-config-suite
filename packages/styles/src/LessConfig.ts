import {BaseConfig, LessLoaderOptions, getLoader} from '@webpack-config-suite/core';
import {LessExtension, CssConfig} from './index';
import {RuleSetUseItem} from 'webpack';

export class LessConfig extends CssConfig {
  public get composed(): typeof BaseConfig[] {
    return [LessExtension];
  }

  protected get test(): RegExp | string {
    return /\.less/;
  }

  protected get lessLoaderOptions(): LessLoaderOptions {
    return null;
  }

  protected get lessLoader(): RuleSetUseItem {
    return getLoader('less-loader', this.lessLoaderOptions);
  }

  protected get use(): RuleSetUseItem[] {
    return super.use.concat(this.lessLoader);
  }
}
