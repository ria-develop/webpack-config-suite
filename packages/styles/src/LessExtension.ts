import {ResolveExtensions} from '@webpack-config-suite/core';

export class LessExtension extends ResolveExtensions {
  protected get extensions(): string[] {
    return ['.less'];
  }
}
