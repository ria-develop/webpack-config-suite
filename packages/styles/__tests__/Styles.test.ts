import {WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration, RuleSetRule, RuleSetUseItem} from 'webpack';
import {CssConfig, LessConfig} from '../src';
import path from 'path';
import {testProcessConfig} from '../../../jest';

const relativePaths = (item: RuleSetUseItem): RuleSetUseItem => {
  if (typeof item === 'object') {
    item.loader = path.relative(process.cwd(), item.loader).replace(/\\/g, '/');
  }
  return item;
};

const fixPaths = (rule: RuleSetRule): RuleSetRule => {
  if (Array.isArray(rule.use)) {
    return {...rule, use: rule.use.map(relativePaths)};
  }
  return rule;
};

describe.each([
  ['development', undefined, {}],
  ['development', undefined, {hot: true}],
  ['development', undefined, {liveReload: true}],
  ['production', undefined, {}]
])(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('given CssConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyCssConfig extends CssConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      testProcessConfig(MyCssConfig, config, env, argv);
      config.module.rules = config.module.rules.map<RuleSetRule>(fixPaths);
      expect(config).toMatchSnapshot();
    });
    it('given LessConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyLessConfig extends LessConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      testProcessConfig(MyLessConfig, config, env, argv);
      config.module.rules = config.module.rules.map<RuleSetRule>(fixPaths);
      expect(config).toMatchSnapshot();
    });
  }
);
