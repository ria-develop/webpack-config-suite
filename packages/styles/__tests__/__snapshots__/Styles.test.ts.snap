// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given CssConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.css/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": true,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "[name].css",
        "filename": "[name].css",
        "ignoreOrder": false,
      },
      "runtimeOptions": Object {
        "attributes": "",
        "insert": "document.head.appendChild(linkTag);",
        "linkType": "text/css",
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given LessConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.less/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": true,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
          "less-loader",
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true}" given CssConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.css/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": undefined,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "[name].css",
        "filename": "[name].css",
        "ignoreOrder": false,
      },
      "runtimeOptions": Object {
        "attributes": "",
        "insert": "document.head.appendChild(linkTag);",
        "linkType": "text/css",
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true}" given LessConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.less/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": undefined,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
          "less-loader",
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given CssConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.css/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": undefined,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "[name].css",
        "filename": "[name].css",
        "ignoreOrder": false,
      },
      "runtimeOptions": Object {
        "attributes": "",
        "insert": "document.head.appendChild(linkTag);",
        "linkType": "text/css",
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given LessConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.less/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": undefined,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
          "less-loader",
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "production", env: "undefined", argv: "{}" given CssConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.css/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": false,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "[name].css",
        "filename": "[name].css",
        "ignoreOrder": false,
      },
      "runtimeOptions": Object {
        "attributes": "",
        "insert": "document.head.appendChild(linkTag);",
        "linkType": "text/css",
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "production", env: "undefined", argv: "{}" given LessConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.less/,
        "use": Array [
          "style-loader",
          Object {
            "loader": "node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": false,
              "hmr": false,
              "reloadAll": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "autoprefixer",
                  ],
                ],
              },
            },
          },
          "less-loader",
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;
